numba/targets/linalg.py

```diff
@@ -667,3 +688,300 @@ def inv(context, builder, sig, args):
+                                        types.intp,  # lda
+                                        types.CPointer(a.dtype),  # wr
+                                        types.CPointer(a.dtype),  # wi
+                                        types.CPointer(a.dtype),  # vl
+                                        types.intp,  # ldvl
+                                        types.CPointer(a.dtype),  # vr
+                                        types.intp  # ldvr
+                                        )
+
+        numba_ez_rgeev = types.ExternalFunction("numba_ez_rgeev",
+                                                numba_ez_rgeev_sig)
+
+        numba_ez_cgeev_sig = types.intc(types.char,  # kind
+                                        types.char,  # jobvl
+                                        types.char,  # jobvr
+                                        types.intp,  # n
+                                        types.CPointer(a.dtype),  # a
+                                        types.intp,  # lda
+                                        types.CPointer(a.dtype),  # w
+                                        types.CPointer(a.dtype),  # vl
+                                        types.intp,  # ldvl
+                                        types.CPointer(a.dtype),  # vr
+                                        types.intp  # ldvr
+                                        )
+
+        numba_ez_cgeev = types.ExternalFunction("numba_ez_cgeev",
+                                                numba_ez_cgeev_sig)
+
+        kind = ord(get_blas_kind(a.dtype, "eig"))
+
+        JOBVL = ord('N')
+        JOBVR = ord('V')
+
+        F_layout = a.layout == 'F'
+
+        def real_eig_impl(a):
+            n = a.shape[-1]
+            if a.shape[-2] != n:
+                msg = "Last 2 dimensions of the array must be square."
+                raise numpy.linalg.LinAlgError(msg)
+
+            if not numpy.isfinite(a).all():
+                raise numpy.linalg.LinAlgError(
+                    "Array must not contain infs or NaNs.")
+
+            if F_layout:
+                acpy = numpy.copy(a)
+            else:
+                acpy = numpy.asfortranarray(a)
+
+            ldvl = 1
+            ldvr = n
+            wr = numpy.zeros(n, dtype=a.dtype)
+            wi = numpy.zeros(n, dtype=a.dtype)
+            vl = numpy.zeros((ldvl, n), dtype=a.dtype)
+            vr = numpy.zeros((ldvr, n), dtype=a.dtype)
+
+            r = numba_ez_rgeev(kind,
+                               JOBVL,
+                               JOBVR,
+                               n,
+                               acpy.ctypes,
+                               n,
+                               wr.ctypes,
+                               wi.ctypes,
+                               vl.ctypes,
+                               ldvl,
+                               vr.ctypes,
+                               ldvr)
+
+            if numpy.any(wi):
```

Comment:
I'm inclined to throw if any non-zero imaginary part is found, regardless of how it may have arisen. If there are genuinely small imaginary parts to the eigenvalues then the user can switch to complex input. If there is numerical issue causing them then by throwing the user finds out. Either way the complex space variant can be used to investigate. If it was not intended for the system to have complex eigenvalues then it says something about the system (depending on what is being modelled, perhaps it is on the cusp of exhibiting oscillatory behaviour?! It also highlights the condition/precision in the data itself). Huge effort has gone into the LAPACK routines to try and produce the best possible answers given what is presented, so I think we just have to trust them. Finally, if users want to ignore the imaginary part, for whatever reason, we have given them a path way to do this.